### ISSUE TITLE
1.2.2: build fails (error: ‘constexpr’ needed for in-class initialization of static data member)

### ISSUE THREAD

**Issue by dvzrv** [sadko4u/lsp-plugins#257]
Hi! I'm currently trying to build 1.2.2 for Arch Linux.
Unfortunately the build fails on many errors all across (an example):

```
mkdir -p /build/lsp-plugins/src/lsp-plugins/.build/target/lsp-dsp-units/main/dynamics/
In file included from /build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/plugins/dyna_processor.h:35,
                 from main/ui/dyna_processor.cpp:22:
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:35:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::THRESHOLD_MIN’ of non-integral type [-fpermissive]
   35 |             static const float  THRESHOLD_MIN           = GAIN_AMP_M_72_DB;
      |                                 ^~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:36:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::THRESHOLD_MAX’ of non-integral type [-fpermissive]
   36 |             static const float  THRESHOLD_MAX           = GAIN_AMP_P_24_DB;
      |                                 ^~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:37:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::THRESHOLD_DFL’ of non-integral type [-fpermissive]
   37 |             static const float  THRESHOLD_DFL           = GAIN_AMP_M_12_DB;
      |                                 ^~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:38:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::THRESHOLD_STEP’ of non-integral type [-fpermissive]
   38 |             static const float  THRESHOLD_STEP          = 0.05f;
      |                                 ^~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:40:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::ATTACK_LVL_MIN’ of non-integral type [-fpermissive]
   40 |             static const float  ATTACK_LVL_MIN          = GAIN_AMP_M_72_DB;
      |                                 ^~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:41:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::ATTACK_LVL_MAX’ of non-integral type [-fpermissive]
   41 |             static const float  ATTACK_LVL_MAX          = GAIN_AMP_P_24_DB;
      |                                 ^~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:42:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::ATTACK_LVL_DFL’ of non-integral type [-fpermissive]
   42 |             static const float  ATTACK_LVL_DFL          = GAIN_AMP_M_12_DB;
      |                                 ^~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:43:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::ATTACK_LVL_STEP’ of non-integral type [-fpermissive]
   43 |             static const float  ATTACK_LVL_STEP         = 0.05f;
      |                                 ^~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:45:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::RELEASE_LVL_MIN’ of non-integral type [-fpermissive]
   45 |             static const float  RELEASE_LVL_MIN         = GAIN_AMP_M_72_DB;
      |                                 ^~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:46:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::RELEASE_LVL_MAX’ of non-integral type [-fpermissive]
   46 |             static const float  RELEASE_LVL_MAX         = GAIN_AMP_P_24_DB;
      |                                 ^~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:47:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::RELEASE_LVL_DFL’ of non-integral type [-fpermissive]
   47 |             static const float  RELEASE_LVL_DFL         = GAIN_AMP_M_12_DB;
      |                                 ^~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:48:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::RELEASE_LVL_STEP’ of non-integral type [-fpermissive]
   48 |             static const float  RELEASE_LVL_STEP        = 0.05f;
      |                                 ^~~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:50:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::ATTACK_TIME_MIN’ of non-integral type [-fpermissive]
   50 |             static const float  ATTACK_TIME_MIN         = 0.0f;
      |                                 ^~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:51:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::ATTACK_TIME_MAX’ of non-integral type [-fpermissive]
   51 |             static const float  ATTACK_TIME_MAX         = 5000.0f;
      |                                 ^~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:52:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::ATTACK_TIME_DFL’ of non-integral type [-fpermissive]
   52 |             static const float  ATTACK_TIME_DFL         = 20.0f;
      |                                 ^~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:53:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::ATTACK_TIME_STEP’ of non-integral type [-fpermissive]
   53 |             static const float  ATTACK_TIME_STEP        = 0.0025f;
      |                                 ^~~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:55:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::RELEASE_TIME_MIN’ of non-integral type [-fpermissive]
   55 |             static const float  RELEASE_TIME_MIN        = 0.0f;
      |                                 ^~~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:56:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::RELEASE_TIME_MAX’ of non-integral type [-fpermissive]
   56 |             static const float  RELEASE_TIME_MAX        = 5000.0f;
      |                                 ^~~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:57:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::RELEASE_TIME_DFL’ of non-integral type [-fpermissive]
   57 |             static const float  RELEASE_TIME_DFL        = 100.0f;
      |                                 ^~~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:58:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::RELEASE_TIME_STEP’ of non-integral type [-fpermissive]
   58 |             static const float  RELEASE_TIME_STEP       = 0.0025f;
      |                                 ^~~~~~~~~~~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:60:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::KNEE_MIN’ of non-integral type [-fpermissive]
   60 |             static const float  KNEE_MIN                = GAIN_AMP_M_24_DB;
      |                                 ^~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:61:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::KNEE_MAX’ of non-integral type [-fpermissive]
   61 |             static const float  KNEE_MAX                = GAIN_AMP_0_DB;
      |                                 ^~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:62:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::KNEE_DFL’ of non-integral type [-fpermissive]
   62 |             static const float  KNEE_DFL                = GAIN_AMP_M_6_DB;
      |                                 ^~~~~~~~
/build/lsp-plugins/src/lsp-plugins/modules/lsp-plugins-dyna-processor/include/private/meta/dyna_processor.h:63:33: error: ‘constexpr’ needed for in-class initialization of static data member ‘const float lsp::meta::dyna_processor_metadata::KNEE_STEP’ of non-integral type [-fpermissive]
   63 |             static const float  KNEE_STEP               = 0.01f;
      |                                 ^~~~~~~~~
```

This is with gcc 12.1 and glibc 2.35.

Full build log: 
[lsp-plugins-1.2.2-1-x86_64-build.log](https://github.com/sadko4u/lsp-plugins/files/8965806/lsp-plugins-1.2.2-1-x86_64-build.log)

**Comment by dvzrv** [comment]
It appears all of those occurences need to be decorated with `constexpr`. I can look into a PR if you want...

**Comment by sadko4u** [comment]
You need to switch to C++98 at this moment. All will be built just fine.

**Comment by dvzrv** [comment]
> You need to switch to C++98 at this moment. All will be built just fine.

Hm, but this was not required with 1.2.1. What changed?

**Comment by sadko4u** [comment]
As @falktTX asked me, LSP Plugins now take CXXFLAGS and CFLAGS from the environment if they are defined. Previously the configuration script has defined them explicitly. You can unset the CXXFLAGS and CFLAGS and then `make config` will set up the proper configuration.
This is also related to https://github.com/sadko4u/lsp-plugins/issues/258

**Comment by dvzrv** [comment]
I have now added `-std=c++98` to CXXFLAGS, but that does not solve the issue I'm seeing in #258 (I don't think that's related to e.g. #253).

**Comment by sadko4u** [comment]
Here is the related commit that 'broke' the usual build.
https://github.com/sadko4u/lsp-plugins/commit/5223ec21740bf3a637bdab4ffe4bb2e4a1e6bcbd

**Comment by dvzrv** [comment]
> You can unset the CXXFLAGS and CFLAGS and then `make config` will set up the proper configuration.

That is not how we build distribution packages unfortunately. The distribution's default flags are there for a reason.

> Here is the related commit that 'broke' the usual build.
> [5223ec2](https://github.com/sadko4u/lsp-plugins/commit/5223ec21740bf3a637bdab4ffe4bb2e4a1e6bcbd)

I don't think it is the best choice to completely replace the project's CFLAGS and CXXFLAGS if the environment provides their own. The externally provided should be appended instead.

**Comment by falkTX** [comment]
> I don't think it is the best choice to completely replace the project's CFLAGS and CXXFLAGS if the environment provides their own. The externally provided should be appended instead.

Yeah, typically if there are flags required by a project they are just appended to the ones from the environment.
So instead of doing `CFLAGS ?= ...` replace it by `CFLAGS += ...` which does the job.

PS: sorry for not have replied in due time to the ticket around this. was going to sometime soon, didnt expect you to release already. but well, now @dvzrv can check on things from a packaging perspective too.

**Comment by sadko4u** [comment]
OK, I will update the configuration script for 1.2.3 then.

**Comment by dvzrv** [comment]
FWIW, the following worked just fine for me:

```
--- a/make/tools.mk     2022-06-23 00:18:02.677549495 +0200
+++ b/make/tools.mk     2022-06-23 13:14:21.572315661 +0200
@@ -110,7 +110,7 @@
 endif

 # Define flags for (cross) build
-CFLAGS             ?= \
+CFLAGS             += \
   $(CFLAGS_EXT) \
   -fdata-sections \
   -ffunction-sections \
@@ -120,7 +120,7 @@

 CDEFS              += -DLSP_INSTALL_PREFIX=\\\"$(PREFIX)\\\"

-CXXFLAGS           ?= \
+CXXFLAGS           += \
   $(CXXFLAGS_EXT) \
   -std=c++98 \
   -fno-exceptions \
```

This also fixes the issues in #258!

**Comment by falkTX** [comment]
note that LDFLAGS from env vars is still not taken into consideration by the build.
this is specially important for LTO builds, which simply wont be enabled here (if requested by distro flags)

from my previous tests, LDFLAGS usage is broken for LSP plugins anyhow, as it invokes `ld` directly instead of the compiler.

we used to call `ld` manually in older days, but the compiler happily does that for us, and better.
because LTO requires the build flags to be present during link stage, it is best to use the compiler for the linking operation instead of the `ld`/linker, as we cannot pass regular build flags to the linker and the compiler can.

**Comment by falkTX** [comment]
FYI I can confirm build is okay, will push to kxstudio repos shortly.
There is a memory issue in http://lsp-plug.in/plugins/lv2/mb_expander_lr due to uninitialized memory, will open a ticket for that later (need to do a debug build in order to see where in the code it happens)

**Comment by sadko4u** [comment]
Building LSP Plugins in debug mode:
```
make clean
make config DEBUG=1
make
```

**Comment by Vascom** [comment]
The same problem at Fedora.

**Comment by sadko4u** [comment]
Refactored code to support C++11 and later.

**Comment by sadko4u** [comment]
Available in 1.2.3 release!
